File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int maxlength = 0;
        Set<Character> charSet = new HashSet<>();
        int l = 0;

        for(int r=0;r<n;r++){
            if(!charSet.contains(s.charAt(r))){
                charSet.add(s.charAt(r));
                maxlength = Math.max(maxlength,r-l+1);
            }
            else{
                while(charSet.contains(s.charAt(r))){
                    charSet.remove(s.charAt(l));
                    l++;
                }
                charSet.add(s.charAt(r));            }
        }
        return maxlength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ch=set()
        count=0
        left=0
        for i in range(len(s)):
            while s[i] in ch:
                ch.remove(s[left])
                left+=1
            ch.add(s[i])
            count = max(count,i-left+1)
        return count